<!DOCTYPE html>
<html>
<head>
  <title>Operational Protein Families</title>
  <meta charset="utf-8">
  <meta name="description" content="Operational Protein Families">
  <meta name="author" content="Kathryn Iverson">
  <meta name="generator" content="slidify" />
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta http-equiv="X-UA-Compatible" content="chrome=1">
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/default.css" media="all" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/phone.css" 
    media="only screen and (max-device-width: 480px)" >
  <link rel="stylesheet" href="libraries/frameworks/io2012/css/slidify.css" >
  <link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" />
  <base target="_blank"> <!-- This amazingness opens all links in a new tab. -->  
  
  <!-- Grab CDN jQuery, fall back to local if offline -->
  <script src="http://ajax.aspnetcdn.com/ajax/jQuery/jquery-1.7.min.js"></script>
  <script>window.jQuery || document.write('<script src="libraries/widgets/quiz/js/jquery.js"><\/script>')</script> 
  <script data-main="libraries/frameworks/io2012/js/slides" 
    src="libraries/frameworks/io2012/js/require-1.0.8.min.js">
  </script>
  
  

</head>
<body style="opacity: 0">
  <slides class="layout-widescreen">
    
    <!-- LOGO SLIDE -->
        <slide class="title-slide segue nobackground">
  <hgroup class="auto-fadein">
    <h1>Operational Protein Families</h1>
    <h2></h2>
    <p>Kathryn Iverson<br/>Bioinformagician</p>
  </hgroup>
  <article></article>  
</slide>
    

    <!-- SLIDES -->
    <slide class="" id="slide-1" style="background:;">
  <hgroup>
    <h2>What:</h2>
  </hgroup>
  <article data-timings="">
    <p>Operational protein families are clusters of similar proteins</p>

<p>Creating OPFs is a database free method of grouping sequences analogous to OTUs</p>

<h2>Why:</h2>

<p>Databases are not great</p>

<ul>
<li>Genes present are limited to what we &quot;know&quot; or at least have seen before</li>
<li>Annotations are limited in the same way</li>
<li>Leaves a large (often over half) of the data in a black box</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="class" id="id" style="background:;">
  <hgroup>
    <h2>How to make OPFs</h2>
  </hgroup>
  <article data-timings="">
    <h3>Quickly:</h3>

<ol>
<li>Assemble</li>
<li>Gene prediction</li>
<li>Get counts</li>
<li>All v all BLAST</li>
<li>Cluster</li>
</ol>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-3" style="background:;">
  <hgroup>
    <h2>Assemble</h2>
  </hgroup>
  <article data-timings="">
    
<div style='float:left;width:48%;'>
  <h3>Iterative velvet</h3>

<ol>
<li>Assemble with a high k value</li>
<li>Enable read tracking and save unused reads</li>
<li>Assemble unused reads at a lower k value</li>
<li>Repeat</li>
</ol>

</div>
<div style='float:right;width:48%;'>
  <h3>Megahit</h3>

<ul>
<li>Runs an iterative assembly by default, small k to large</li>
<li>Much faster than iterative velvet</li>
<li>Yes, I know I&#39;m mixing numbers and bullets.</li>
</ul>

</div>
  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-4" style="background:;">
  <hgroup>
    <h2>Gene prediction</h2>
  </hgroup>
  <article data-timings="">
    <h3>Metagene annotator (MGA)</h3>

<ul>
<li>Gene prediction models for Bacteria, Archaea and Phage</li>
<li>Uses ribosomal binding site patterns as predictors for domain</li>
<li>Optimized for short fragments

<ul>
<li>uses 700bp fragment as example</li>
<li>similar to contig sizes seen after assembly</li>
</ul></li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-5" style="background:;">
  <hgroup>
    <h2>Estimate gene counts</h2>
  </hgroup>
  <article data-timings="">
    <p>Map reads with bowtie for short reads, bwa for long reads and normalize to gene length</p>

<h3>Counts file</h3>

<ul>
<li>Use bowtie to map reads to genes</li>
<li>Normalize by sequence length </li>
</ul>

<pre><code>ceiling(num_reads_mapped * 100 / len(gene))
</code></pre>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-6" style="background:;">
  <hgroup>
    <h2>Creating OPFs</h2>
  </hgroup>
  <article data-timings="">
    <h3>BLAST</h3>

<p>All v all</p>

<p>Blast all genes against each other (proteins, blastp)</p>

<pre><code>-max_target_seqs 10000
</code></pre>

<h3>Cluster</h3>

<p>mg-cluster with mothur</p>

<pre><code>mgcluster(blast=allVall100.out, count=allgenes.preg.counts)
</code></pre>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-7" style="background:;">
  <hgroup>
    <h2>Datasets</h2>
  </hgroup>
  <article data-timings="">
    <ol>
<li>Obesity in twins (Turnbaugh et al. 2009)

<ul>
<li>18 metagenomes from lean and obese twin pairs and their mothers</li>
<li>From MG-RAST</li>
</ul></li>
<li>Pregnancy (Koren et al. 2012)

<ul>
<li>20 metagenomes from 10 mothers during the first and third trimester of pregnancy</li>
<li>From NCBI (MG-RAST data is messed up and being fixed with zero urgency)</li>
</ul></li>
<li>HMP data

<ul>
<li>138 metagenomes</li>
</ul></li>
</ol>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-8" style="background:;">
  <hgroup>
    <h2>HMP dataset</h2>
  </hgroup>
  <article data-timings="">
    <p>Dataset was too large to cluster as a whole so OPFs were created within the KEGG category to which the genes were a best match</p>

<ol>
<li>BLAST genes against KEGG</li>
<li>Group genes by their KEGG top hit</li>
<li>Cluster seperatly within this group</li>
</ol>

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-9" style="background:;">
  <hgroup>
    <h2>Twin data</h2>
  </hgroup>
  <article data-timings="">
    
<div style='float:left;width:48%;'>
  <p><img src="assets/fig/unnamed-chunk-2-1.png" alt="plot of chunk unnamed-chunk-2"> </p>

</div>
<div style='float:right;width:48%;'>
  <ul>
<li>Turnbaugh et al argue there is no taxanomic core but there may be a functional core</li>
<li>This seems to be true with the very coarse metric of KEGG functional categories</li>
<li>OPFs reflect the 16s (dis)similarity</li>
</ul>

</div>
  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-10" style="background:;">
  <article data-timings="">
    <p><img src="assets/img/figure_2.svg" alt="image"></p>

<!-- this file went missing before this was under version control. Need to find it or re-generate it...but this is the code to make the plot

Figure 2

t.pcoa <- read.table(file='twin_pcoa_opf_kegg_16s_vars.csv', sep=',', header=T)

ggplot(t.pcoa, aes(axis1, axis2, color=weight, shape=as.factor(group), group=as.factor(family))) + geom_point(size=4) + theme_bw() + xlab("Dimension 1") + ylab("Dimension 2") + geom_polygon(aes(mapping=group, alpha=1)) + facet_wrap(~method, scales='free', ncol=1) + theme(plot.background = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_blank(), panel.background = element_blank()) + theme(axis.line = element_line(color = 'black'))

-->

  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-11" style="background:;">
  <hgroup>
    <h2>Pregnancy data</h2>
  </hgroup>
  <article data-timings="">
    
<div style='float:left;width:48%;'>
  <p><img src="assets/fig/unnamed-chunk-3-1.png" alt="plot of chunk unnamed-chunk-3"> </p>

</div>
<div style='float:right;width:48%;'>
  <ul>
<li>Distances are very small with the KEGG pcoa</li>
<li>Similar to the twin dataset, OPFs tend to follow OTUs</li>
<li>Functional core may be a reflection of database limitations </li>
</ul>

</div>
  </article>
  <!-- Presenter Notes -->
</slide>

<slide class="" id="slide-12" style="background:;">
  <article data-timings="">
    <!--
 
## I'm a commented slide 1
 
From Slidify's authoring process, this slide was made:
1. Edit YAML front matter
2. Write using R Markdown
3. Use an empty line followed by three dashes to separate slides!

![plot of chunk unnamed-chunk-4](assets/fig/unnamed-chunk-4-1.png) 
 
-->

<h2>Annotation</h2>

<ul>
<li>BLAST genes to a database of your choice, KEGG for example, and annotate OPF by majority vote.</li>
<li>They&#39;re still only as good as your database though...</li>
</ul>

  </article>
  <!-- Presenter Notes -->
</slide>

    <slide class="backdrop"></slide>
  </slides>
  <div class="pagination pagination-small" id='io2012-ptoc' style="display:none;">
    <ul>
      <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=1 title='What:'>
         1
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=2 title='How to make OPFs'>
         2
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=3 title='Assemble'>
         3
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=4 title='Gene prediction'>
         4
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=5 title='Estimate gene counts'>
         5
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=6 title='Creating OPFs'>
         6
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=7 title='Datasets'>
         7
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=8 title='HMP dataset'>
         8
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=9 title='Twin data'>
         9
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=10 title=''>
         10
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=11 title='Pregnancy data'>
         11
      </a>
    </li>
    <li>
      <a href="#" target="_self" rel='tooltip' 
        data-slide=12 title=''>
         12
      </a>
    </li>
  </ul>
  </div>  <!--[if IE]>
    <script 
      src="http://ajax.googleapis.com/ajax/libs/chrome-frame/1/CFInstall.min.js">  
    </script>
    <script>CFInstall.check({mode: 'overlay'});</script>
  <![endif]-->
</body>
  <!-- Load Javascripts for Widgets -->
  
  <!-- LOAD HIGHLIGHTER JS FILES -->
  <script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
  <script>hljs.initHighlightingOnLoad();</script>
  <!-- DONE LOADING HIGHLIGHTER JS FILES -->
   
  </html>